cmd/heartbeats/main.go

```diff
@@ -43,13 +50,37 @@ func init() {
 	flag.StringVar(&periodStr, "period", "5", "the number of seconds between heartbeats")
 }
 
+type envConfig struct {
+	// Sink URL where to send heartbeat cloudevents
+	Sink string `envconfig:"SINK"`
+
+	// Name of this pod.
+	Name string `envconfig:"POD_NAME"`
+
+	// Namespace this pod exists in.
+	Namespace string `envconfig:"POD_NAMESPACE"`
+
+	// The ip address of this pod.
+	IP string `envconfig:"POD_IP"`
+}
+
 func main() {
 	flag.Parse()
 
-	client := cloudevents.NewClient(sink, cloudevents.Builder{
-		EventType: "dev.knative.source.heartbeats",
-		Source:    "heartbeats-demo",
-	})
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Printf("[ERROR] Failed to process env var: %s", err)
+		os.Exit(1)
+	}
+
+	if env.Sink != "" {
```

Comment:
Just a question if the Process above will actually fail if POD_NAME, POD_NAMESPACE are not specified? Since Sink can be empty, can Name / Namespace be too? it seems like they should fail if not specified?  Should you tag them with required?

Reply:
it will not, they will just default to `""` but that will not be possable using the sample yaml.

I am trying to show that the source url changes per instance of heartbeat.

Reply:
marked required.

Reply:
Hard to understand from the documents if that's the way you specify two things. I trust you it works :)